### DISCUSSION TITLE
How to get rid of KHR_Texture_Transform extension?

### DISCUSSION THREAD

**Discussion by lokiiarora** [donmccurdy/glTF-Transform#1095] [Q&A] [answered]
Hey,
basically I'm trying to build an atlas solution in our 3D Asset Pipeline, where I have figured out a way to put all of it back into an atlas of sorts using naive packing techniques but then one road block I have is `KHR_Texture_Transform` and there is no way for me to remove that extension and sort of put its scale and offset data onto the `TEXCOORD_*` attribute itself.

Also, because of KHR_Texture_Transform, meshopt also refuses to apply it's compression on the given attribute / primitive as seen from the error logs

**Comment by donmccurdy** [comment]
It's certainly possible to modify vertex data in glTF Transform, by reading an element (e.g. the of UV coordinates for a vertex) from an [Accessor](https://gltf-transform.dev/modules/core/classes/Accessor), applying the transform, and writing the result back to the accessor. 

You may need to be careful to understand _why_ KHR_texture_transform is there, though... e.g. if different materials use the same mesh data with different texture transforms, then you'd need to make copies of the mesh data before applying changes to the UVs.

Do I understand correctly what you're trying to do?

***

> Also, because of KHR_Texture_Transform, meshopt also refuses to apply it's compression on the given attribute / primitive as seen from the error logs

I'm not aware of any such constraint. Are you using glTF Transform to do meshopt compression, or something else? What error do you see in the logs?

If you're able to share a model related to these problems, that almost always saves time figuring out the solution.

**Comment by donmccurdy** [comment, marked as answer]
To remove `KHR_texture_transform` and apply its transforms to each mesh primitive:

```typescript
import { Material } from '@gltf-transform/core';
import { Transform } from '@gltf-transform/extensions';
import { prune } from '@gltf-transform/functions';
import { vec3, mat3 } from 'gl-matrix';

const matrixMap = new Map();

// Compute UV transform matrix for each material's base color texture.
for (const material of document.getRoot().listMaterials()) {
	const info = material.getBaseColorTextureInfo();
	const transform = info.getExtension('KHR_texture_transform');
	if (!transform) continue;
	
	const matrix = mat3.create();
	mat3.translate(matrix, matrix, transform.getOffset());
	mat3.rotate(matrix, matrix, -transform.getRotation());
	mat3.scale(matrix, matrix, transform.getScale());	

	matrixMap.set(material, matrix);
}

// Transform UVs.
for (const mesh of document.getRoot().listMeshes()) {
	for (const prim of mesh.listPrimitives()) {
		const matrix = matrixMap.get(prim.getMaterial());
		if (!matrix) continue;
		
		const uv = prim.getAttribute('TEXCOORD_0').clone();
		for (let i = 0, el = [0, 0, 1], il = uv.getCount(); i < il; i++) {
			uv.getElement(i, el);
			vec3.transformMat3(el, el, matrix);
			uv.setElement(i, el);
		}
		prim.setAttribute('TEXCOORD_0', uv);
	}
}

// Remove KHR_texture_transform extension.
for (const extension of document.getRoot().listExtensionsUsed()) {
	if (extension.extensionName === 'KHR_texture_transform') {
		extension.dispose();
	}
}

// Clean up.
await document.transform(prune());
```

This code makes a number of assumptions and simplifications:

1. `TEXCOORD_0` attributes are float32
2. we modify `baseColorTexture` and leave all other textures alone
3. `baseColorTexture` always uses `TEXCOORD_0`
4. each primitive uses all vertices in its vertex buffer, or, we don't have a situation where different primitives use the same vertex buffers with different indices

When processing models that don't meet those assumptions, or arbitrary models from unknown users, adjustments would need to be made.
